rpc/src/v1/helpers/light_fetch.rs

```diff
@@ -367,8 +368,42 @@ impl LightFetch {
 							}
 							future::ok::<_,OnDemandError>(matches)
 						}) // and then collect them into a vector.
-						.map(|matches| matches.into_iter().map(|(_, v)| v).collect())
 						.map_err(errors::on_demand_error)
+						// retrieve transaction hash.
+						.and_then(move |matches| {
+							let mut blocks = BTreeMap::new();
+							let mut result: Vec<Log> = matches.into_iter().map(|(_, v)| {
+								{
+								let block_hash = v.block_hash.as_ref().expect("Previously initialized with value; qed");
+								blocks.entry(block_hash.clone()).or_insert_with(|| {
+									fetcher_block.block(BlockId::Hash(block_hash.clone().into()))
+								});
+								}
+								v
+							}).collect();
+							// future get blocks (unordered it)
+							stream::futures_unordered(blocks.into_iter().map(|(_,v)|v)).collect().map(move |blocks| {
+								let mut tr_per_bl = BTreeMap::new();
```

Comment:
`tree_per_block`? Seems to store all blocks as <Hash, Transactions>

And why is this not a HashMap? BTreeMap performs linear search O(n)?